lib/bolt/transport/orch.rb

```diff
@@ -58,36 +64,122 @@ def run_script(target, script, arguments, _options = {})
+          # the result otherwise make sure an error is generated
+          if state == 'finished' || (result && result['_error'])
+            Bolt::Result.new(target, value: result)
+          elsif state == 'skipped'
+            Bolt::Result.new(
+              target,
+              value: { '_error' => {
+                'kind' => 'puppetlabs.tasks/skipped-node',
+                'msg' => "Node #{target.host} was skipped",
+                'details' => {}
+              } }
+            )
+          else
+            # Make a generic error with a unkown exit_code
+            Bolt::Result.for_task(target, result.to_json, '', 'unknown')
+          end
+        end
+      end
+
+      def batch_command(targets, command, _options = {})
+        promises = batch_task(targets,
+                              BOLT_MOCK_FILE,
+                              'stdin',
+                              action: 'command',
+                              command: command,
+                              options: {})
+        promises.map { |promise| promise.then { |result| unwrap_bolt_result(result.target, result) } }
+      end
+
+      def batch_script(targets, script, arguments, _options = {})
+        content = File.open(script, &:read)
+        content = Base64.encode64(content)
+        params = {
+          action: 'script',
+          content: content,
+          arguments: arguments
+        }
+        promises = batch_task(targets, BOLT_MOCK_FILE, 'stdin', params)
+        promises.map { |promise| promise.then { |result| unwrap_bolt_result(result.target, result) } }
+      end
+
+      def batch_upload(targets, source, destination, _options = {})
+        content = File.open(source, &:read)
+        content = Base64.encode64(content)
+        mode = File.stat(source).mode
+        params = {
+          action: 'upload',
+          path: destination,
+          content: content,
+          mode: mode
+        }
+        promises = batch_task(targets, BOLT_MOCK_FILE, 'stdin', params)
+        promises.map do |promise|
+          promise.then do |result|
+            result.error_hash ? result : Bolt::Result.for_upload(result.target, source, destination)
+          end
+        end
+      end
+
+      def batch_task(targets, task, _inputmethod, arguments, _options = {})
+        callback = block_given? ? Proc.new : proc {}
+        targets.group_by { |target| target.options[:orch_task_environment] }.flat_map do |_environment, env_targets|
+          body = build_request(env_targets, task, arguments)
+
+          # Make promises to hold the eventual results. The `set` operation is
+          # fast, so they can just execute in the thread that's creating them,
+          # using the :immediate executor.
+          result_ivars = Hash[env_targets.map { |target| [target, Concurrent::Promise.new(executor: :immediate)] }]
+
+          @executor.post do
+            env_targets.each do |target|
+              callback.call(type: :node_start, target: target)
```

Comment:
We get a real start event from orch. There should at least be a TODO to use that here.